File: macos/Tokendial/HoverCard.swift
import AppKit
import TokendialCore

/// The detail card under a hovered cell: every window with its reset copy, a bar, both ends of the number, then the live sessions.
enum HoverCard {
    static func build(_ tile: Tile, now: Date) -> NSView {
        let body = NSStackView()
        body.orientation = .vertical
        body.alignment = .leading
        body.spacing = 0
        body.translatesAutoresizingMaskIntoConstraints = false

        let header = row(left: Label.make(tile.name, size: 14, color: Theme.textPrimary, weight: .semibold), right: tile.account.map { Label.make($0.summary, size: 11, color: Theme.textSecondary, alignment: .right) })
        add(header, to: body, top: 0)

        switch tile.reading.status {
        case .needsSignIn: add(wrap(Label.make("Not signed in.", size: 12, color: Theme.textSecondary)), to: body, top: 6)
        case .unsupported(let why): add(wrap(Label.make(why, size: 12, color: Theme.textSecondary)), to: body, top: 6)
        case .failed(let why): add(wrap(Label.make("Could not read usage (\(why)). Showing nothing rather than a guess.", size: 12, color: Theme.textSecondary)), to: body, top: 6)
        default: break
        }

        for window in tile.reading.windows {
            let reset = window.resetsAt.map { Copy.reset($0, now: now) } ?? ""
            add(row(left: Label.make(window.label, size: 12, color: Theme.textPrimary), right: Label.make(reset, size: 11, color: Theme.textSecondary, alignment: .right)), to: body, top: 8)
            if let fraction = window.usedFraction {
                let bar = BarView(height: 4)
                bar.set(fraction, color: Theme.color(tile.reading.block != nil ? .critical : Band.of(fraction)))
                add(bar, to: body, top: 5)
            }
            add(Label.make(window.summary(tile.reading.fidelity), size: 11, color: Theme.textSecondary), to: body, top: 4)
        }

        if let block = tile.reading.block {
            add(Label.make(Copy.until(block.reason, block.until, now: now), size: 11, color: Theme.critical), to: body, top: 8)
        }
        if case .stale(let since) = tile.reading.status, since > .distantPast {
            add(Label.make("Last read \(Copy.ago(since, now: now))", size: 11, color: Theme.textDisabled), to: body, top: 8)
        }
        if tile.reading.fidelity == .derived {
            add(wrap(Label.make("Derived from local activity, not published by the provider", size: 11, color: Theme.textDisabled)), to: body, top: 6)
        }
        if let activity = tile.activity, !activity.sessions.isEmpty {
            let line = NSView()
            line.wantsLayer = true
            line.layer?.backgroundColor = Theme.hairline.cgColor
            line.translatesAutoresizingMaskIntoConstraints = false
            line.heightAnchor.constraint(equalToConstant: 1).isActive = true
            add(line, to: body, top: 10)
            for session in activity.ordered.prefix(4) { add(sessionRow(session, now: now), to: body, top: 6) }
        }

        let card = NSView()
        card.wantsLayer = true
        card.layer?.backgroundColor = Theme.cardSurface.cgColor
        card.layer?.cornerRadius = Theme.cardRadius
        card.layer?.borderWidth = 1
        card.layer?.borderColor = Theme.surfaceEdge.cgColor
        card.translatesAutoresizingMaskIntoConstraints = false
        card.addSubview(body)
        NSLayoutConstraint.activate([
            card.widthAnchor.constraint(equalToConstant: Theme.cardWidth),
            body.leadingAnchor.constraint(equalTo: card.leadingAnchor, constant: Theme.cardPadding),
            body.trailingAnchor.constraint(equalTo: card.trailingAnchor, constant: -(Theme.cardPadding + 2)),
            body.topAnchor.constraint(equalTo: card.topAnchor, constant: Theme.cardPadding),
            body.bottomAnchor.constraint(equalTo: card.bottomAnchor, constant: -Theme.cardPadding)
        ])
        return card
    }

    private static func add(_ view: NSView, to stack: NSStackView, top: CGFloat) {
        stack.addView(view, in: .bottom)
        view.widthAnchor.constraint(equalTo: stack.widthAnchor).isActive = true
        if top > 0 { stack.setCustomSpacing(top, after: stack.views[max(0, stack.views.count - 2)]) }
    }

    private static func row(left: NSView, right: NSView?) -> NSView {
        let row = NSStackView()
        row.orientation = .horizontal
        row.distribution = .fill
        row.spacing = 8
        row.translatesAutoresizingMaskIntoConstraints = false
        row.addView(left, in: .leading)
        if let right {
            right.setContentCompressionResistancePriority(.required, for: .horizontal)
            right.setContentHuggingPriority(.required, for: .horizontal)
            if let field = right as? NSTextField { field.preferredMaxLayoutWidth = 130 }
            row.addView(right, in: .trailing)
        }
        return row
    }

    private static func wrap(_ field: NSTextField) -> NSTextField {
        field.lineBreakMode = .byWordWrapping
        field.maximumNumberOfLines = 0
        field.preferredMaxLayoutWidth = Theme.cardWidth - 2 * Theme.cardPadding - 2
        return field
    }

    static func sessionRow(_ session: AgentSession, now: Date) -> NSView {
        let row = NSStackView()
        row.orientation = .horizontal
        row.spacing = 8
        row.alignment = .centerY
        row.translatesAutoresizingMaskIntoConstraints = false
        let dot = NSView()
        dot.wantsLayer = true
        dot.layer?.cornerRadius = 3
        dot.layer?.backgroundColor = (session.state == .waiting ? Theme.waiting : session.state == .working ? Theme.working : Theme.textDisabled).cgColor
        dot.translatesAutoresizingMaskIntoConstraints = false
        dot.widthAnchor.constraint(equalToConstant: 6).isActive = true
        dot.heightAnchor.constraint(equalToConstant: 6).isActive = true
        row.addView(dot, in: .leading)
        let label = session.state == .waiting && session.waitingFor != nil ? "\(session.name) · \(session.waitingFor!)" : "\(session.name) · \(session.location)"
        row.addView(Label.make(label, size: 12, color: Theme.textPrimary), in: .leading)
        let when = Label.make(session.state == .waiting ? "waiting \(Copy.elapsed(session.since, now: now))" : Copy.elapsed(session.since, now: now), size: 11, color: Theme.textSecondary, alignment: .right)
        when.setContentCompressionResistancePriority(.required, for: .horizontal)
        row.addView(when, in: .trailing)
        return row
    }
}

File: windows/Tokendial.App/Panel/HoverCard.cs
using System.Windows;
using System.Windows.Controls;
using Tokendial.Core.I18n;
using Tokendial.Core.Model;
using Tokendial.Core.Sessions;

namespace Tokendial.App.Panel;

/// <summary>The detail card under a hovered cell: every window with its reset copy, a bar, both ends of the number, then the live sessions.</summary>
public static class HoverCard
{
    public static Border Build(Tile tile, DateTimeOffset now)
    {
        var body = new StackPanel { HorizontalAlignment = HorizontalAlignment.Stretch };
        var header = new DockPanel { LastChildFill = true, Margin = new Thickness(0, 0, 0, 6) };
        var mark = new MarkView(tile.Id, 16) { Fill = Theme.TextPrimary, VerticalAlignment = VerticalAlignment.Center, Margin = new Thickness(0, 0, 8, 0) };
        DockPanel.SetDock(mark, Dock.Left);
        header.Children.Add(mark);
        var title = Text.Primary(tile.Name, 14, FontWeights.SemiBold);
        var account = tile.Account?.Summary;
        if (!string.IsNullOrEmpty(account))
        {
            var right = Text.Secondary(account, 11, TextAlignment.Right);
            right.MaxWidth = 130;
            right.Margin = new Thickness(8, 0, 0, 0);
            DockPanel.SetDock(right, Dock.Right);
            header.Children.Add(right);
        }
        header.Children.Add(title);
        body.Children.Add(header);

        switch (tile.Reading.Status)
        {
            case ReadingStatus.NeedsSignIn:
                body.Children.Add(Wrap(Text.Secondary(Strings.T("card.notSignedIn"), 12)));
                break;
            case ReadingStatus.Unsupported u:
                body.Children.Add(Wrap(Text.Secondary(u.Why, 12)));
                break;
            case ReadingStatus.Failed f:
                body.Children.Add(Wrap(Text.Secondary(Strings.T("card.failed", ("why", f.Why)), 12)));
                break;
        }

        foreach (var window in tile.Reading.Windows)
        {
            var row = new DockPanel { LastChildFill = true, Margin = new Thickness(0, 6, 0, 0) };
            var reset = window.ResetsAt is DateTimeOffset at ? Copy.Reset(at, now) : "";
            var resetBlock = Text.Secondary(reset, 11, TextAlignment.Right);
            resetBlock.Margin = new Thickness(8, 0, 0, 0);
            DockPanel.SetDock(resetBlock, Dock.Right);
            row.Children.Add(resetBlock);
            row.Children.Add(Text.Primary(Strings.Label(window.Label), 12));
            body.Children.Add(row);
            if (window.UsedFraction is double fraction)
            {
                body.Children.Add(WithMargin(Text.Bar(fraction, 4, Theme.Of(tile.Reading.Block is not null ? Band.Critical : Bands.Of(fraction))), new Thickness(0, 5, 0, 4)));
            }
            body.Children.Add(Text.Secondary(window.Summary(tile.Reading.Fidelity), 11));
            if (tile.Forecast is Forecast forecast && window.Id == tile.Headline?.Id)
            {
                var pace = Text.Make(Copy.Forecast(forecast, now), 11, forecast.Kind == ForecastKind.RunsOut ? Theme.Watch : Theme.TextSecondary);
                pace.TextWrapping = System.Windows.TextWrapping.Wrap;
                pace.TextTrimming = System.Windows.TextTrimming.None;
                body.Children.Add(WithMargin(pace, new Thickness(0, 2, 0, 0)));
            }
        }

        if (tile.Reading.Block is Blocked block)
        {
            body.Children.Add(WithMargin(Text.Make(Copy.Until(block.Reason, block.Until, now), 11, Theme.Critical), new Thickness(0, 8, 0, 0)));
        }

        if (tile.Reading.Status is ReadingStatus.Stale stale && stale.Since > DateTimeOffset.MinValue)
        {
            body.Children.Add(WithMargin(Text.Disabled(Strings.T("card.lastRead", ("ago", Copy.Ago(stale.Since, now))), 11), new Thickness(0, 8, 0, 0)));
        }
        if (tile.Reading.Fidelity == Fidelity.Derived)
        {
            body.Children.Add(WithMargin(Text.Disabled(Strings.T("card.derived"), 11), new Thickness(0, 6, 0, 0)));
        }

        if (tile.Activity is Activity activity && activity.Sessions.Count > 0)
        {
            body.Children.Add(Text.Hairline(new Thickness(0, 10, 0, 8)));
            foreach (var session in activity.Ordered.Take(4)) body.Children.Add(SessionRow(session, now));
        }

        return new Border
        {
            Width = Theme.CardWidth,
            Background = Theme.CardSurface,
            BorderBrush = Theme.SurfaceEdge,
            BorderThickness = new Thickness(1),
            CornerRadius = new CornerRadius(Theme.CardRadius),
            Padding = new Thickness(Theme.CardPadding, Theme.CardPadding, Theme.CardPadding + 2, Theme.CardPadding),
            Child = body,
            UseLayoutRounding = false
        };
    }

    public static DockPanel SessionRow(AgentSession session, DateTimeOffset now)
    {
        var row = new DockPanel { LastChildFill = true, Margin = new Thickness(0, 2, 0, 2) };
        var dot = new Border { Width = 6, Height = 6, CornerRadius = new CornerRadius(3), Margin = new Thickness(0, 0, 8, 0), VerticalAlignment = VerticalAlignment.Center,
            Background = session.State == SessionState.Waiting ? Theme.Waiting : session.State == SessionState.Working ? Theme.Working : Theme.TextDisabled };
        DockPanel.SetDock(dot, Dock.Left);
        row.Children.Add(dot);
        var when = Text.Secondary(session.State == SessionState.Waiting ? Strings.T("card.waiting", ("elapsed", Copy.Elapsed(session.Since, now))) : Copy.Elapsed(session.Since, now), 11, TextAlignment.Right);
        when.Margin = new Thickness(8, 0, 0, 0);
        DockPanel.SetDock(when, Dock.Right);
        row.Children.Add(when);
        var label = session.WaitingFor is string why && session.State == SessionState.Waiting ? $"{session.Name} · {why}" : $"{session.Name} · {session.Where}";
        row.Children.Add(Text.Primary(label, 12));
        return row;
    }

    private static TextBlock Wrap(TextBlock block)
    {
        block.TextWrapping = TextWrapping.Wrap;
        block.TextTrimming = TextTrimming.None;
        return block;
    }

    private static T WithMargin<T>(T element, Thickness margin) where T : FrameworkElement
    {
        element.Margin = margin;
        return element;
    }
}
